pkg/agent/containerd/containerd.go

```diff
@@ -186,36 +184,111 @@ func preloadImages(ctx context.Context, cfg *config.Node) error {
+		images, err = prePullImages(ctx, client, imageClient, file)
+		if err != nil {
+			return errors.Wrap(err, "failed to pull images from "+filePath)
+		}
+	} else {
+		opener, err := tarfile.GetOpener(filePath)
+		if err != nil {
+			return err
+		}
+
+		imageReader, err := opener()
+		if err != nil {
+			return err
+		}
+		defer imageReader.Close()
+
+		logrus.Infof("Importing images from %s", filePath)
+		images, err = client.Import(ctx, imageReader, containerd.WithAllPlatforms(true))
+		if err != nil {
+			return errors.Wrap(err, "failed to import images from "+filePath)
+		}
 	}
 
-	opener, err := tarfile.GetOpener(filePath)
-	if err != nil {
-		return err
+	if err := labelImages(ctx, client, images); err != nil {
+		return errors.Wrap(err, "failed to add pinned label to images")
+	}
+	if err := retagImages(ctx, client, images, cfg.AgentConfig.AirgapExtraRegistry); err != nil {
+		return errors.Wrap(err, "failed to retag images")
 	}
 
-	imageReader, err := opener()
+	for _, image := range images {
+		logrus.Infof("Imported %s", image.Name)
+	}
+	return nil
+}
+
+// clearLeases deletes any leases left by previous versions of k3s.
+// We no longer use leases to lock content; they only locked the
+// blobs, not the actual images.
+func clearLeases(ctx context.Context, client *containerd.Client) error {
+	ls := client.LeasesService()
+	existingLeases, err := ls.List(ctx)
 	if err != nil {
 		return err
 	}
-	defer imageReader.Close()
-
-	logrus.Infof("Importing images from %s", filePath)
+	for _, lease := range existingLeases {
+		if lease.ID == version.Program {
+			logrus.Debugf("Deleting existing lease: %v", lease)
+			ls.Delete(ctx, lease)
+		}
+	}
+	return nil
+}
 
-	images, err := client.Import(ctx, imageReader, containerd.WithAllPlatforms(true))
+// clearLabels removes the pinned labels on all images in the image store that were previously pinned by k3s
+func clearLabels(ctx context.Context, client *containerd.Client) error {
+	var errs []error
+	imageService := client.ImageService()
+	images, err := imageService.List(ctx, fmt.Sprintf("labels.%q==%s", k3sPinnedImageLabelKey, k3sPinnedImageLabelValue))
 	if err != nil {
 		return err
 	}
+	for _, image := range images {
+		delete(image.Labels, k3sPinnedImageLabelKey)
+		delete(image.Labels, labels.PinnedImageLabelKey)
+		if _, err := imageService.Update(ctx, image, "labels"); err != nil {
+			errs = append(errs, errors.Wrap(err, "failed to delete labels from image "+image.Name))
+		}
+	}
+	return merr.NewErrors(errs...)
+}
 
-	return retagImages(ctx, client, images, cfg.AgentConfig.AirgapExtraRegistry)
+// labelImages adds labels to the listed images, indicating that they
+// are pinned by k3s and should not be pruned.
+func labelImages(ctx context.Context, client *containerd.Client, images []images.Image) error {
+	var errs []error
+	imageService := client.ImageService()
+	for i, image := range images {
+		if image.Labels[k3sPinnedImageLabelKey] == k3sPinnedImageLabelValue &&
```

Comment:
sorry small comment here, is there a chance that image.Labels may be nil, in that case we can expect a panic here

Reply:
No. You can check keys in a nil map without panicing. You just can't assign to a nil map.